File: GFG/BST/kthsmall.cpp
#include<iostream>
#include<vector>
using namespace std;

class KNode{
public:
    KNode(int x) : key(x), left(nullptr), right(nullptr), lcount(0) {}
    int key;
    KNode* left;
    KNode* right;
    int lcount;
    KNode* insert(KNode* node,int data);
    void inorder(KNode* root);
};

void insert(KNode* node, int data){
    if(node==NULL){
        KNode* treod = new KNode(data);
        node=treod;
        return ;
    }
    else{
        if(node->key>data){
            if(node->left==NULL){
                node->lcount++;
                KNode* treod = new KNode(data);
                node->left=treod;
            }
            else{
                node->lcount++;
                insert(node->left,data);
            }
        }
        else{
            if(node->right==NULL){
                KNode* treod = new KNode(data);
                node->right=treod;
            }
            else{
                insert(node->right,data);
            }
        }
    }
    return ;
}
void inorder(KNode* root){
    if(root==NULL){
        return;
    }
    inorder(root->left);
    cout<<root->key<<" * "<<root->lcount<<endl;
    inorder(root->right);
}

int main(){
    vector<int> arr {20,100,10,40,70,120,4,60,80};
    KNode* root=new KNode(50);

    for (auto i:arr){
        insert(root,i);
    }

    inorder(root);
    return 0;
}

File: GFG/BST/kthsmall.py
import ds

class KNode:
    def __init__(self,data):
        self.key=data
        self.left=None
        self.right=None
        self.lcount=0
        
    def __insert(self,node,data):  # sourcery skip: remove-redundant-pass
        if data < node.key:
            if node.left is None:
                node.lcount+=1
                node.left = KNode(data)
            else:
                node.lcount+=1
                self.__insert(node.left, data)
        elif data > node.key:
            if node.right is None:
                node.right = KNode(data)
            else:
                self.__insert(node.right, data)
        else:
            # Value already exists in the tree, handle accordingly
            pass
    def insert(self,node,data):
        if node is None:
            node = KNode(data)
        else:
            self.__insert(node,data)
            
    def inorder(self,node):
        if(node is None):
            return
        self.inorder(node.left)
        print(node.key," * ",node.lcount)
        self.inorder(node.right)
        
arr=[20,100,10,40,70,120,60,80,4]
root=KNode(50)
for i in arr:
    root.insert(root,i)

root.inorder(root)
        